ext-apiserver/pkg/controller/machineset/controller.go

```diff
@@ -29,24 +32,110 @@ import (
 type MachineSetControllerImpl struct {
 	builders.DefaultControllerFns
 
-	// lister indexes properties about MachineSet
-	lister listers.MachineSetLister
+	// machineClient a client that knows how to consume Machine resources
+	machineClient machineclientset.Interface
+
+	// machineSetsLister indexes properties about MachineSet
+	machineSetsLister listers.MachineSetLister
+
+	// machineLister holds a lister that knows how to list Machines from a cache
+	machineLister listers.MachineLister
 }
 
 // Init initializes the controller and is called by the generated code
 // Register watches for additional resource types here.
 func (c *MachineSetControllerImpl) Init(arguments sharedinformers.ControllerInitArguments) {
-	// Use the lister for indexing machinesets labels
-	c.lister = arguments.GetSharedInformers().Factory.Cluster().V1alpha1().MachineSets().Lister()
+	c.machineSetsLister = arguments.GetSharedInformers().Factory.Cluster().V1alpha1().MachineSets().Lister()
+	c.machineLister = arguments.GetSharedInformers().Factory.Cluster().V1alpha1().Machines().Lister()
+
+	var err error
+	c.machineClient, err = machineclientset.NewForConfig(arguments.GetRestConfig())
+	if err != nil {
+		glog.Fatalf("error building clientset for machineClient: %v", err)
+	}
 }
 
-// Reconcile handles enqueued messages
-func (c *MachineSetControllerImpl) Reconcile(u *v1alpha1.MachineSet) error {
-	// TODO: Implement controller logic here
-	glog.Infof("Running reconcile MachineSet for %s\n", u.Name)
-	return nil
+// Reconcile holds the controller's business logic.
+// it makes sure that the current state is equal to the desired state.
+// note that the current state of the cluster is calculated based on the number of machines
+// that are owned by the given machineSet (key).
+func (c *MachineSetControllerImpl) Reconcile(machineSet *v1alpha1.MachineSet) error {
+	filteredMachines, err := c.getMachines(machineSet)
+	if err != nil {
+		return err
+	}
+
+	return c.syncReplicas(machineSet, filteredMachines)
 }
 
 func (c *MachineSetControllerImpl) Get(namespace, name string) (*v1alpha1.MachineSet, error) {
-	return c.lister.MachineSets(namespace).Get(name)
+	return c.machineSetsLister.MachineSets(namespace).Get(name)
+}
+
+// syncReplicas essentially scales machine resources up and down.
+func (c *MachineSetControllerImpl) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
+	var result error
+	currentMachineCount := int32(len(machines))
+	desiredReplicas := *machineSet.Spec.Replicas
+	diff := int(currentMachineCount - desiredReplicas)
+
+	if diff < 0 {
```

Comment:
at the moment resync is set to 10 minutes thus I decided to create/delete resources in a loop so that the results are visible more quickly.

Reply:
I think that in the future we could also add watch events on machines - that would "wake up" appropriate replica.